waku/node/v2/message_store.nim

```diff
@@ -0,0 +1,266 @@
+  if x.val != nil:
+    dispose(x.release)
+
+template checkErr(op, cleanup: untyped) =
+  if (let v = (op); v != SQLITE_OK):
+    cleanup
+    return err($sqlite3_errstr(v))
+
+template checkErr(op) =
+  checkErr(op): discard
+
+proc init*(
+    T: type MessageStore,
+    basePath: string,
+    name: string = "store",
+    readOnly = false,
+    inMemory = false): MessageStoreResult[T] =
+  var env: AutoDisposed[ptr sqlite3]
+  defer: disposeIfUnreleased(env)
+
+  let
+    name =
+      if inMemory: ":memory:"
+      else: basepath / name & ".sqlite3"
+    flags =
+      if readOnly: SQLITE_OPEN_READONLY
+      else: SQLITE_OPEN_READWRITE or SQLITE_OPEN_CREATE
+
+  if not inMemory:
+    try:
+      createDir(basePath)
+    except OSError, IOError:
+      return err("`sqlite: cannot create database directory")
+
+  checkErr sqlite3_open_v2(name, addr env.val, flags.cint, nil)
+
+  template prepare(q: string, cleanup: untyped): ptr sqlite3_stmt =
+    var s: ptr sqlite3_stmt
+    checkErr sqlite3_prepare_v2(env.val, q, q.len.cint, addr s, nil):
+      cleanup
+    s
+
+  template checkExec(s: ptr sqlite3_stmt) =
+    if (let x = sqlite3_step(s); x != SQLITE_DONE):
+      discard sqlite3_finalize(s)
+      return err($sqlite3_errstr(x))
+
+    if (let x = sqlite3_finalize(s); x != SQLITE_OK):
+      return err($sqlite3_errstr(x))
+
+  template checkExec(q: string) =
+    let s = prepare(q): discard
+    checkExec(s)
+
+  template checkWalPragmaResult(journalModePragma: ptr sqlite3_stmt) =
+    if (let x = sqlite3_step(journalModePragma); x != SQLITE_ROW):
+      discard sqlite3_finalize(journalModePragma)
+      return err($sqlite3_errstr(x))
+
+    if (let x = sqlite3_column_type(journalModePragma, 0); x != SQLITE3_TEXT):
+      discard sqlite3_finalize(journalModePragma)
+      return err($sqlite3_errstr(x))
+
+    if (let x = sqlite3_column_text(journalModePragma, 0);
+        x != "memory" and x != "wal"):
+      discard sqlite3_finalize(journalModePragma)
+      return err("Invalid pragma result: " & $x)
+
+  # TODO: check current version and implement schema versioning
+  checkExec "PRAGMA user_version = 1;"
+
+  let journalModePragma = prepare("PRAGMA journal_mode = WAL;"): discard
+  checkWalPragmaResult(journalModePragma)
+  checkExec(journalModePragma)
+
+  ## Table is the SQL query for creating the messages Table.
+  ## It contains:
+  ##  - 4-Byte ContentTopic stored as an Integer
+  ##  - Payload stored as a blob
+  checkExec """
+    CREATE TABLE IF NOT EXISTS messages (
+        id BLOB PRIMARY KEY,
+        timestamp INTEGER NOT NULL,
+        contentTopic INTEGER NOT NULL, 
+        payload BLOB
+    ) WITHOUT ROWID;
+    """
+
+  ok(MessageStore(
+    env: env.release
+  ))
+
+template prepare(env: Sqlite, q: string, cleanup: untyped): ptr sqlite3_stmt =
+  var s: ptr sqlite3_stmt
+  checkErr sqlite3_prepare_v2(env, q, q.len.cint, addr s, nil):
+    cleanup
+  s
+
+proc bindParam(s: RawStmtPtr, n: int, val: auto): cint =
```

Comment:
I have no idea what this is, and I'm not confident in our ability to address bugs if things break here. I'd much rather this be upstream, as right now it looks like magic that we don't really understand. (I'm sure it makes sense to people in the Nimbus team).

Ditto a bunch of the templates above.